client/wait.go

```diff
@@ -0,0 +1,248 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/zrepl/zrepl/cli"
+	"github.com/zrepl/zrepl/config"
+	"github.com/zrepl/zrepl/daemon"
+	"github.com/zrepl/zrepl/daemon/job"
+	"github.com/zrepl/zrepl/daemon/pruner"
+	"github.com/zrepl/zrepl/daemon/snapper"
+	"github.com/zrepl/zrepl/replication/report"
+)
+
+type wui struct {
+	//lock      sync.Mutex //For report and error
+	report map[string]*job.Status
+
+	jobName string
+}
+
+var WaitCmd = &cli.Subcommand{
+	Use:   "wait JOB",
+	Short: "block and wait until the specified job is done or has failed",
+	Run: func(ctx context.Context, subcommand *cli.Subcommand, args []string) error {
+		return runWaitCmd(subcommand.Config(), args)
+	},
+}
+
+// adapted from status.go
+func runWaitCmd(config *config.Config, args []string) error {
+	if len(args) != 1 {
+		return errors.Errorf("Expected 1 argument: JOB")
+	}
+
+	// TODO: poll status until done or error
+	httpc, err := controlHttpClient(config.Global.Control.SockPath)
+	if err != nil {
+		return err
+	}
+
+	t := wui{
+		report:  nil,
+		jobName: args[0],
+	}
+
+	updateAndEvaluate := func() (bool, error) {
+		var m daemon.Status
+
+		err := jsonRequestResponse(httpc, daemon.ControlJobEndpointStatus,
+			struct{}{},
+			&m,
+		)
+		if err != nil {
+			return true, fmt.Errorf("can't get status from daemon: %w", err)
+		}
+
+		t.report = m.Jobs
+
+		return t.evaluate()
+	}
+	updateAndEvaluate()
+
+	ticker := time.NewTicker(500 * time.Millisecond)
+	defer ticker.Stop()
+ticker:
+	for range ticker.C {
+		done, err := updateAndEvaluate()
+		if err != nil {
+			return err
+		}
+		if done {
+			break ticker
+		}
+	}
+
+	return nil
+}
+
+func (t *wui) evaluate() (bool, error) {
+	v, ok := t.report[t.jobName]
+	if !ok {
+		return true, fmt.Errorf("job not in status: %v", t.jobName)
+	}
+
+	if v.Type == job.TypePush || v.Type == job.TypePull {
+		activeStatus, ok := v.JobSpecific.(*job.ActiveSideStatus)
+		if !ok || activeStatus == nil {
+			// TODO: ignore?
+			return true, nil //error.New("ActiveSideStatus is null")
+		}
+
+		//t.renderReplicationReport(activeStatus.Replication, t.getReplicationProgressHistory(k))
+		replicationDone, err := evaluateReplicationReport(activeStatus.Replication)
+		if err != nil {
+			return true, err
+		}
+
+		senderPruningDone, err := evaluatePruningReport(activeStatus.PruningSender)
+		if err != nil {
+			return true, fmt.Errorf("sender %w", err)
+		}
+		receiverPruningDone, err := evaluatePruningReport(activeStatus.PruningReceiver)
+		if err != nil {
+			return true, fmt.Errorf("receiver %w", err)
+		}
+
+		done := replicationDone && senderPruningDone && receiverPruningDone
+
+		if v.Type == job.TypePush {
+			snapperDone, err := evaluateSnapperReport(activeStatus.Snapshotting)
+			if err != nil {
+				return true, err
+			}
+
+			done = done && snapperDone
```

Comment:
Handling the snapshotting is tricky: snapshotting is never exactly 'done'. Maybe the command should be more like `zrepl wait-job JOBNAME --replication [true,false] --snapshotting [true,false] --pruning [true,false]` ?

See first commit in status-v2 `add signal 'snapshot', rename existing signal 'wakeup' to 'replication'` [link](https://github.com/zrepl/zrepl/pull/297/commits/a1a0c9f30552dbe2d42fd19a88e4dd40f88d8b57)

Reply:
hmmm ... I was wondering if there're situations where a job finished fast enough that the wait after a signal would be stuck waiting on a job thinking that it's not started yet.

Reply:
I'd suggest keeping it simple at first and see if there's actually a need for sub-job granularity.